### PR TITLE
NoticeTemplateReporter: Add an option for treating projects like packages

### PR DESCRIPTION
**Pull request by fviernau**
ORT has two different notice reporter templates. The notice summary
template just lists all licenses in one large list. The default
template has one joint list for the licenses of all projects and
separate lists for the licenses of each package.

In some cases it makes sense to also list the licenses separetly for (some
of the) projects. In particular, listing the licenses for
SpdxDocumentFile projects separately is interresting, because this
allows to create a notice by package report also in cases where no
package manager is involved and the source code of the project dependencies
is just included in the source tree of the analyzed project.

In order to support that use case the scan results have to be
deduplicated for the projects which are to be reported separately in the
notices, because the scan results of projects may overlap - which would
in turn result in licenses being reported twice, in the list for the
respective package and in the joint licenses list for the projects.

### CODE REVIEW

**Comment by sschuberth** [comment]
> in cases where no
> package manager is involved and the source code of the project dependencies
> is just included in the source tree of the analyzed project.

Yes, that is a general problem that we only have with the `SpdxDocumentFile` package manager, and I was giving this some thought recently, too.

Our [spdx-project-expected-output.yml](https://github.com/oss-review-toolkit/ort/blob/master/analyzer/src/funTest/assets/projects/synthetic/spdx-project-expected-output.yml) nicely highlights the problem in the following lines.

Project VCS path:
https://github.com/oss-review-toolkit/ort/blob/c3a2d02b9141c68911f65b9ee46ebb0d30ed627f/analyzer/src/funTest/assets/projects/synthetic/spdx-project-expected-output.yml#L18

Package "curl" VCS path:
https://github.com/oss-review-toolkit/ort/blob/c3a2d02b9141c68911f65b9ee46ebb0d30ed627f/analyzer/src/funTest/assets/projects/synthetic/spdx-project-expected-output.yml#L56

Package "openssl" VCS path:
https://github.com/oss-review-toolkit/ort/blob/c3a2d02b9141c68911f65b9ee46ebb0d30ed627f/analyzer/src/funTest/assets/projects/synthetic/spdx-project-expected-output.yml#L87

So scanning the project path would scan the files from the packages, too, and associate the findings to the project! Also, scanning the packages separately would scan the files belonging to packages again. This obviously is wrong. But I believe it should be fixed on the scanner level, not on the reporter level. IMO this is yet another use-case for the scan-by-repo topic and the related filtering: Project scan results should include the findings from below their parent directory, but exclude and findings from nested packages.

**Comment by fviernau** [comment]
> his obviously is wrong. But I believe it should be fixed on the scanner level, not on the reporter level. IMO this is yet another use-case for the scan-by-repo topic and the related filtering: Project scan results should include the findings from below their parent directory, but exclude and findings from nested packages.

I fully agree with the first path until the colon. I tried to implement this outside of the reporter in the first place but in each attempt I run into issues where what I was trying to accomplish did not fit into the design and thus wasn't possible without large refactoring.
The only reason I'm proposing this PR as-is now is because I need a working solution anytime soon.
I've kept the implementation quite isolated so that it won't be a hazzle to remove it again as soon as we have scan by repository done.

Apart from scan by repository which is already a large effort there is (as you mentioned) also the topic of how to associate files to projects which I believe isn't low effort either. @sschuberth  Given that would you be ok to go with the implementation I propose here until the larger blockers are resolved?

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt [hunk 1]
```diff
@@ -54,13 +60,41 @@ class FreemarkerTemplateProcessor(
...
      * generated files.
      */
     fun processTemplates(input: ReporterInput, outputDir: File, options: Map<String, String>): List<File> {
+        val projectTypesAsPackages = options[OPTION_PROJECT_TYPES_AS_PACKAGES]?.split(",").orEmpty().toSet()
+        val projectsAsPackages = input.ortResult.getProjects().map { it.id }.filterTo(mutableSetOf()) {
+            it.type in projectTypesAsPackages
+        }
+
+        if (projectTypesAsPackages.isNotEmpty()) {
+            log.info { "Handling ${projectTypesAsPackages.size} of types $projectTypesAsPackages as packages." }
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:77]: Add "projects" after `${projectTypesAsPackages.size}`.

File: model/src/main/kotlin/licenses/LicenseInfoResolver.kt [hunk 2]
```diff
@@ -44,7 +44,7 @@ class LicenseInfoResolver(
     val provider: LicenseInfoProvider,
     val copyrightGarbage: CopyrightGarbage,
     val archiver: FileArchiver?,
-    licenseFilenamePatterns: LicenseFilenamePatterns = LicenseFilenamePatterns.DEFAULT
+    val licenseFilenamePatterns: LicenseFilenamePatterns = LicenseFilenamePatterns.DEFAULT
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/licenses/LicenseInfoResolver.kt:47]: Commit message: Typos in "separetly", "interresting"

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt [hunk 3]
```diff
@@ -239,3 +274,106 @@ private fun List<ResolvedLicense>.merge(): ResolvedLicense {
         locations = flatMapTo(mutableSetOf()) { it.locations }
     )
 }
+
+/**
+ * Return an [OrtResult] with all license and copyright findings associated with [targetProjects] removed from all other
+ * projects not in [targetProjects].
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:280]: Please emphasize that this affects projects which have a target project in a subdirectory, this would help understanding what the function does.

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt [hunk 4]
```diff
@@ -54,13 +60,41 @@ class FreemarkerTemplateProcessor(
...
+            options = options,
+            projectsAsPackages = projectsAsPackages
+        )
+    }
+
+    /**
+     * Process all Freemarker templates referenced in "template.id" and "template.path" options and returns the
+     * generated files.
+     */
+    fun processTemplatesInternal(
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:92]: Should this have the `internal` modifier?

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt [hunk 5]
```diff
@@ -54,13 +60,41 @@ class FreemarkerTemplateProcessor(
...
+        val projectsAsPackages = input.ortResult.getProjects().map { it.id }.filterTo(mutableSetOf()) {
+            it.type in projectTypesAsPackages
+        }
+
+        if (projectTypesAsPackages.isNotEmpty()) {
+            log.info { "Handling ${projectTypesAsPackages.size} of types $projectTypesAsPackages as packages." }
+        }
+
+        return processTemplatesInternal(
+            input = input.deduplicateProjectScanResults(projectsAsPackages),
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:81]: I think this feature is complex and could be difficult to debug, a test case would be much appreciated.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt (hunk 5 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:81]: I'd propose to add a test case which verifies that the `findings` related to the `target project types` do get `subtracted` from the scan result of `all other` projects. That is, the test case would verify the scan results contained in the `OrtResult`.

Would that test case address what you are missing @mnonnenmacher ?

Comment (mnonnenmacher) [reply reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:81]: I think testing this on the level of `processTemplates()` would be more complete, but a test case for `deduplicateProjectScanResults()` would probably be more elegant. Both would be fine for me as most of the complexity is in the code of `deduplicateProjectScanResults()`.

Comment (fviernau) [reply reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:84]: Done.

File: reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+import org.ossreviewtoolkit.model.config.ScannerConfiguration
+import org.ossreviewtoolkit.utils.test.DEFAULT_ANALYZER_CONFIGURATION
+
+private fun scanResultContainer(
+    id: String,
+    vcsInfo: VcsInfo,
+    findingsPaths: Collection<String>
+): ScanResultContainer {
+    val licenseFindings = findingsPaths.mapTo(sortedSetOf()) { LicenseFinding("MIT", TextLocation(it, 1)) }
+    val copyrightFindings = findingsPaths.mapTo(sortedSetOf()) { CopyrightFinding("MIT", TextLocation(it, 1)) }
```
Comment (mnonnenmacher) [inline reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt:59]: Better use something like "(c)" here instead of "MIT".

File: reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+        val ortResult = ORT_RESULT.deduplicateProjectScanResults(targetProjects)
+
+        "keep the findings of all target projects" {
+            targetProjects.forAll { targetProject ->
+                ortResult.getScanResultsForId(targetProject) shouldBe ORT_RESULT.getScanResultsForId(targetProject)
+            }
+        }
+
+        "filter the findings of the root project as expected" {
```
Comment (mnonnenmacher) [inline reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt:173]: More explicit: "remove the findings of all target projects from the root project"

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt (hunk 6 above)
Comment (fviernau) [reply reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt:59]: Done.

File: reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt (hunk 7 above)
Comment (fviernau) [reply reporter/src/test/kotlin/utils/FreeMarkerTemplateProcessorTest.kt:173]: Done.

**Review by mnonnenmacher — APPROVED** [review]

**Comment by sschuberth** [comment]
> Given that would you be ok to go with the implementation I propose here until the larger blockers are resolved?

I must admit that I'm still not happy with the change, esp. when looking at the amount of code / complexity it adds. Moreover, I believe we recently tend to make too many "I need to get this working now"-style of changes without fixing the underlying architecture and questioning past design choices. We need to get back to taking that "luxury" to avoid running into the technical debt / legacy code hell.

TL;DR I won't block the change.

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt [hunk 8]
```diff
@@ -54,13 +61,43 @@ class FreemarkerTemplateProcessor(
     companion object {
         const val OPTION_TEMPLATE_ID = "template.id"
         const val OPTION_TEMPLATE_PATH = "template.path"
+        const val OPTION_PROJECT_TYPES_AS_PACKAGES = "project.types.as.packages"
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt:64]: I completely missed that property name before, dots should be used to create a hierarchy, like with "template.id" and "template.path". This should be called "project-types-as-packages" instead. Also, this new option must be added to the class docs of all reports that use the `FreemarkerTemplateProcessor`.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Comment by fviernau** [comment]
> > Given that would you be ok to go with the implementation I propose here until the larger blockers are resolved?
> 
> I must admit that I'm still not happy with the change, esp. when looking at the amount of code / complexity it adds. Moreover, I believe we recently tend to make too many "I need to get this working now"-style of changes without fixing the underlying architecture and questioning past design choices. We need to get back to taking that "luxury" to avoid running into the technical debt / legacy code hell.
> 
> TL;DR I won't block the change.

You are aware that we commited to doing the scan by repository story and the needed file archiver changes as next topic.
So this will infact implement the properly designed solution. Thus, we are actually taking that "luxury". Just not now now.

File: reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt (hunk 8 above)
Comment (fviernau) [reply reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt]: Done.

**Comment by mnonnenmacher** [comment]
> > Given that would you be ok to go with the implementation I propose here until the larger blockers are resolved?
> 
> I must admit that I'm still not happy with the change, esp. when looking at the amount of code / complexity it adds. Moreover, I believe we recently tend to make too many "I need to get this working now"-style of changes without fixing the underlying architecture and questioning past design choices. We need to get back to taking that "luxury" to avoid running into the technical debt / legacy code hell.
> 
> TL;DR I won't block the change.

@sschuberth @fviernau How I understand your previous discussion this feature can probably be removed again once scan by repository is implemented? If that is correct, can you create a ticket for this before merging the PR to make sure we do not forget about it?

@sschuberth To add some background, this is the last step of a long story that @fviernau is working on and for which he also removed some serious technical debt like the inconsistent configuration of file patterns. I would suggest that we put the topic of how to handle the situation when external dependencies are part of the source tree on the agenda of the next developer meeting.

@nnobelis @fviernau Could you please check if this change already helps with the problem that https://github.com/oss-review-toolkit/ort/pull/3508 tries to fix?

**Review by mnonnenmacher — APPROVED** [review]

**Comment by fviernau** [comment]
> @sschuberth @fviernau How I understand your previous discussion this feature can probably be removed again once scan by repository is implemented? If that is correct, can you create a ticket for this before merging the PR to make sure we do not forget about it?

Answering this in very detail would require to have a very exact understanding of what changes are being considered in scope of the "scan by repository" epic. I don't think it's that clear yet, but let me try to elaborate on the question anyway: 

If we should change how findings / files get associated with projects and if we should conclude a non-pverlapping is the way to go, then the de-duplication could be removed from that feature which would simplify it a lot. But, I believe we still need an option to tell ORT that we do want separate entries for all projects of a specific type (e.g. SpdxDocumentFile) in the NOTICE file.

> @nnobelis @fviernau Could you please check if this change already helps with the problem that #3508 tries to fix?

I'm not certain whether #3508 was targeting only the effect on the NOTICE file.
If it wasn't targeting the NOTICE file this PR would not help.
If it was targeting the NOTICE file then this PR would not help again, because the change proposed in 3508 would, IIUC, eliminate all findings for these SpdxDocumentFile projects as opposed to adding them as separate entry / list in the NOTICES file.

Anyhow, if it was then the proposed change would remove the SpdxDocumentFile package full from the NOTICES while this change doesn't.
